### PR TITLE
Only add card once

### LINKED ISSUE #19: Only add Road and City events once
RIght now it is possible to to add the same Road/City event over and over again

### PR DESCRIPTION
**Pull request by signed**
This fixes #19;

One question:
A `CardStack` has a `list` and a `stack` property. City and Road events are added to the stack while Item Designs and Single Items are added to the list.

What is the difference between list and stack in `CardStack`?

### CODE REVIEW

**Comment by kosta** [comment]
Stack is what you draw from; It is ordered because you can add cards "from the bottom".
List is a bunch of visible cards.

Most of the time, only one is used (e.g. Road & City events only use the Stack) but not always: e.g. random items have a stack of cards to draw and a list of cards that were drawn.

File: src/app.tsx [hunk 1]
```diff
@@ -0,0 +1,779 @@
...
+
+  constructor(private item: ItemUrl, readonly offset: number) {
+    this.item = item;
+    this.url = item.url;
+    this.offset = offset;
+    this.n = item.n;
+  }
+}
+
+type CardRenderProps = RandomItemDesignProps | ItemProps | PersonalGoalProps | RandomSideScenarioProps;
```
Comment (kosta) [inline src/app.tsx:56]: # 👍

File: src/app.tsx [hunk 2]
```diff
@@ -0,0 +1,779 @@
...
+    maxItem: 150,
+    url: "https://lh3.googleusercontent.com/u/0/d/1BHfEwqmC_dax5dV4RFFP5MlfJ76eS7KZ=s3200-k-iv2",
+  }];
+
+  let acc = 0;
+  let itemSheet = 0;
+  return itemCounts.map((num, i) => {
+    if (i >= 71 && i <= 95) {
+      // there are two random items but only one of them is "red", i.e. in this picture
+      num -= 1;
```
Comment (kosta) [inline src/app.tsx:100]: I can still feel the pain of the night when I wrote this comment ;)

File: src/app.tsx [hunk 3]
```diff
@@ -0,0 +1,779 @@
...
+      }, this.save);
+    } else {
+      // assuming "City Event" or "Road Event"
+      const eventStackName = cardType.split(" ")[0].toLowerCase() + "Events";
+      const stack = this.state.stacks[eventStackName];
+      if (!stack || typeof stack === "number") {
+        throw "Unknown name for addCards: " + cardType;
+      }
+
+      const notAlreadyContainedCards = unique(cardIdsToAdd).filter((it) => !stack.stack.includes(it));
```
Comment (kosta) [inline src/app.tsx:690]: Is this the only change? Moving from index.tsx to app.tsx AND doing code changes was a good way to hide the actual few lines of changes ;)

**Review by kosta — COMMENTED** [review]
Looks good, but almost impossible to review...

File: src/app.tsx (hunk 2 above)
Comment (signed) [reply src/app.tsx:100]: :)

File: src/app.tsx (hunk 3 above)
Comment (signed) [reply src/app.tsx:690]: I agree, I should have made two separate pull requests. Initially the fragment change was also part of this on  :)

Comment (signed) [reply src/app.tsx:690]: And yes, this is what changed.

**Comment by signed** [comment]
Any changes you want to have in this PR?

**Comment by kosta** [comment]
# ❤️

**Pull request merged**